File: src/v3/CryptChatV2/Services/NgrokService.cs
using System.Text.Json;

namespace CryptChatV2.Services
{
    public class NgrokService
    {
        private const string NgrokExecutablePath = @"C:\ProgramData\chocolatey\bin\ngrok.exe";
        private const string NgrokApiUrl = "http://localhost:4040/api/tunnels";

        public static async Task<string> GetNgrokUrlAsync()
        {
            var process = new System.Diagnostics.Process
            {
                StartInfo = new System.Diagnostics.ProcessStartInfo
                {
                    FileName = NgrokExecutablePath,
                    Arguments = "http https://localhost:7013"
                }
            };
            process.Start();
            await Task.Delay(5000);

            using var client = new HttpClient();
            var ngrokStatusResponse = await client.GetStringAsync(NgrokApiUrl);

            using (JsonDocument doc = JsonDocument.Parse(ngrokStatusResponse))
            {
                return doc.RootElement
                    .GetProperty("tunnels")[0]
                    .GetProperty("public_url")
                    .GetString();
            }
        }
    }
}

File: src/v3/CryptChatV2.F#/Services/NgrokService.fs
namespace CryptChat.Services

open System
open System.Diagnostics
open System.Net.Http
open System.Text.Json

module NgrokService =
    let private ngrokExecutablePath = @"C:\ProgramData\chocolatey\bin\ngrok.exe"
    let private ngrokApiUrl = "http://localhost:4040/api/tunnels"

    let getNgrokUrlAsync () =
        async {
            let processStartInfo = ProcessStartInfo(FileName = ngrokExecutablePath, Arguments = "http https://localhost:5000")
            use process = new Process(StartInfo = processStartInfo)
            process.Start() |> ignore

            do! Async.Sleep 5000

            use client = new HttpClient()
            let! ngrokStatusResponse = client.GetStringAsync(ngrokApiUrl) |> Async.AwaitTask

            use doc = JsonDocument.Parse(ngrokStatusResponse)
            let publicUrl = 
                (doc.RootElement.GetProperty("tunnels")).[0]

            let result = publicUrl.GetProperty("public_url").GetString()

            return result
        }